### PR TITLE
Route level config support for AWS request signing filter

### PR DESCRIPTION
**Pull request by PeterL328**
Commit Message: Add route-level config for AWS request signing filter.
Additional Description: If route-level config is specified then it overrides the configuration on the virtual host level if set. Route-level sets a stats prefix so more granular stats may be exposed.
Risk Level: Low
Testing: Unit tests
Docs Changes: included
Release Notes: N/A
Platform Specific Features: N/A
[Optional Runtime guard:]
[Optional Fixes #Issue] #21541
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by PeterL328** [comment]
cc: @JuniorHsu

File: test/extensions/filters/http/aws_request_signing/config_test.cc
Comment (JuniorHsu) [inline test/extensions/filters/http/aws_request_signing/config_test.cc:22]: move the above two alias to config.h so we can benefits more

File: api/envoy/extensions/filters/http/aws_request_signing/v3/aws_request_signing.proto
Comment (JuniorHsu) [inline api/envoy/extensions/filters/http/aws_request_signing/v3/aws_request_signing.proto:70]: we can drop v2 support here.

**Review by JuniorHsu — CHANGES_REQUESTED** [review]
Request changes for fixing ci failures.
We might need to have filter tests as well but I'll let code owners to decide.

**Comment by PeterL328** [comment]
@lizan Want to get some early feedback from you regarding proposed API changes :). 
Let me what you think of having the per-route message using shared `AwsRequestSigning` message which may lock us in there but seems fine to me.

File: api/envoy/extensions/filters/http/aws_request_signing/v3/aws_request_signing.proto
Comment (lizan) [inline api/envoy/extensions/filters/http/aws_request_signing/v3/aws_request_signing.proto:69]: From the impl, this seems to override the whole message, so a clarification here would be helpful (i.e. this is not merging at field level in `AwsRequestSigning`).

File: source/extensions/filters/http/aws_request_signing/aws_request_signing_filter.cc
Comment (lizan) [inline source/extensions/filters/http/aws_request_signing/aws_request_signing_filter.cc:102]: Why this need to be a non-const?

File: api/envoy/extensions/filters/http/aws_request_signing/v3/aws_request_signing.proto
Comment (PeterL328) [reply api/envoy/extensions/filters/http/aws_request_signing/v3/aws_request_signing.proto:69]: Wonder if merging at field level is better than override the entire message?
If we decide to merge at field level we will need to add each of the fields from AwsRequestSigning to AwsRequestSigningPerRoute and may need to maintain two messages here.

Thoughts?

Comment (lizan) [reply api/envoy/extensions/filters/http/aws_request_signing/v3/aws_request_signing.proto:69]: Not saying merging field level is better or not, it depends on the possible use cases. Overriding a whole message generally is less confusing, but if you expect most case just overriding one field (e.g. `hostname`), then a field level merging could avoid repetitive config in every route, especially when other fields like `match_excluded_headers` is large.

File: test/extensions/filters/http/aws_request_signing/config_test.cc
Comment (suniltheta) [inline test/extensions/filters/http/aws_request_signing/config_test.cc:54]: Do we want to include `host_rewrite` for line coverage?

File: api/envoy/extensions/filters/http/aws_request_signing/v3/aws_request_signing.proto
Comment (PeterL328) [reply api/envoy/extensions/filters/http/aws_request_signing/v3/aws_request_signing.proto:69]: Yeah, I think `match_excluded_headers` could get large and possible we just want to override on a few selected fields. 
Leaning towards modifying my changes to support field level override.

File: source/extensions/filters/http/aws_request_signing/aws_request_signing_filter.cc
Comment (PeterL328) [reply source/extensions/filters/http/aws_request_signing/aws_request_signing_filter.cc:102]: Hey lizan,
It needs to be non-const because after getting the config using above method we have calls to non-const member methods on config such as [.stats()](https://github.com/envoyproxy/envoy/blob/3914a92fb2d086d450dcf4cc3f0d2571a52c60f0/source/extensions/filters/http/aws_request_signing/aws_request_signing_filter.cc#L57) and .signer().

File: test/extensions/filters/http/aws_request_signing/config_test.cc
Comment (PeterL328) [reply test/extensions/filters/http/aws_request_signing/config_test.cc:54]: yea good idea.

File: docs/root/configuration/http/http_filters/aws_request_signing_filter.rst
Comment (PeterL328) [reply docs/root/configuration/http/http_filters/aws_request_signing_filter.rst:55]: Sounds good. Will make sure proto changes are good with API reviewer and will make necessary adjusted if needed there.

Will clean-up doc and use bootstrap yaml

Comment (PeterL328) [reply docs/root/configuration/http/http_filters/aws_request_signing_filter.rst:62]: Moved to bootstrap yaml

File: api/envoy/extensions/filters/http/aws_request_signing/v3/aws_request_signing.proto
Comment (PeterL328) [reply api/envoy/extensions/filters/http/aws_request_signing/v3/aws_request_signing.proto:69]: Added clarification

File: test/extensions/filters/http/aws_request_signing/config_test.cc
Comment (PeterL328) [reply test/extensions/filters/http/aws_request_signing/config_test.cc:56]: Updated

**Comment by phlax** [comment]
@PeterL328 looks like you need to fix DCO, but happy to do a docs review

/docs

**Comment by PeterL328** [comment]
> @PeterL328 looks like you need to fix DCO, but happy to do a docs review
> 
> /docs
@phlax Thanks :)
I cleaned up the docs. Let me know what you think of the docs.

**Comment by phlax** [comment]
```console
Code coverage for source/extensions/filters/http/aws_request_signing is lower than limit of 96.6 (96.3)
```

/coverage

File: source/extensions/filters/http/aws_request_signing/config.cc
Comment (suniltheta) [inline source/extensions/filters/http/aws_request_signing/config.cc:40]: can replace `envoy::extensions::filters::http::aws_request_signing::v3::AwsRequestSigningPerRoute` with `AwsRequestSigningProtoPerRouteConfig` now?

Comment (PeterL328) [reply source/extensions/filters/http/aws_request_signing/config.cc:40]: Yes, Good catch

**Comment by PeterL328** [comment]
Looks like CI passed. 
Let me know what you think about the API changes :) @lizan

**Review by suniltheta — DISMISSED** [review]
Thanks for this change. Do you want to mention Fixes: https://github.com/envoyproxy/envoy/issues/21541 in the overview?

**Comment by PeterL328** [comment]
> Thanks for this change. Do you want to mention Fixes: #21541 in the overview?

Thanks,
I've added the issue number in the description above.

**Comment by PeterL328** [comment]
@lizan Need one more approval from API-reviewer. 🙂

**Comment by PeterL328** [comment]
@phlax Do you mind taking a look😁

File: docs/root/configuration/http/http_filters/aws_request_signing_filter.rst
Comment (phlax) [inline docs/root/configuration/http/http_filters/aws_request_signing_filter.rst:53]:
```suggestion
Above shows an example of route-level config overriding the config on the virtual-host level.
```

**Review by phlax — COMMENTED** [review]
lgtm from docs pov, thanks @PeterL328 

modulo one small typo

**Comment by PeterL328** [comment]
Rebased with main

**Comment by phlax** [comment]
@adisuissa would you mind doing an api review

**Review by lizan — APPROVED** [review]

**Comment by PeterL328** [comment]
Thanks everyone. 
Looks like ready to merge @mattklein123

**Comment by PeterL328** [comment]
@phlax  Wonder if you are able to merge?

**Pull request merged**
